### PR TITLE
Rework line drafts page

### PR DESCRIPTION
**Pull request by HenrikHartiala**
LineDraftsPage will now show all the draft routes which are under the given line label. This is mostly for MVP and this page will be modified in the future to have much more details about route versions etc.

There is also two commits to:
* Decouple useGetLineDetails from RouteStopsSection, so that RouteStopsSection is reusable.
* Move setObservationDate (to query parameters) to useUrlQuery hook, so that it is also reusable.

---
This change is [Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)

### CODE REVIEW

**Review by tk-vincit — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 15 files reviewed, 3 unresolved discussions (waiting on @HenrikHartiala and @tk-vincit)

___
*[`ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx` line 30 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SpzsYEQHgXxwW9qJ6:-N8SpzsYEQHgXxwW9qJ7:bhw76ky) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx#L30)):*
> ```XML
>   };
> 
>   /** Determines and sets date to query parameters if it's not there */
> ```

Why do we have to set "initial" observation date to query? IMO it is valid case that it is not there (and then we would fallback to e.g. today in the UI)
___
*[`ui/src/hooks/useUrlQuery.ts` line 6 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SqxSZ8JPV6Thv_GSy:-N8SqxSZ8JPV6Thv_GSz:b-aby0kk) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/useUrlQuery.ts#L6)):*
> ```XML
> import { useHistory, useLocation } from 'react-router-dom';
> 
> export const useUrlQuery = () => {
> ```

Do we want to expand responsibilites of this hook or rather create new hook to do new things?

Seems like we could split this easily to 2 different hooks that would do just one thing each:
- `useUrlQuery()` (or readUrlQuery or parseUrlQuery or something) to do what this originally did
- `setToUrlQuery(newParam: Object)`: calls useUrlQuery and sets new query param based to it

Then we could have convenience hooks for setting different parts of url query (e.g. observationDate) to avoid having define same (string) names for parts of url query in different places. And then you could expose all the previous hooks from one general hook, so that UI components could access all of those from one import and from one useSomeHok call.
___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 56 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SsjKbBMDChXhi0Vha:-N8SsjKbBMDChXhi0Vhb:b-o65tbq) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L56)):*
> ```XML
>   const routes = mapRouteResultToRoutes(result);
> 
>   // Filter routes by observationDate in front end (and not in gql query) to avoid
> ```

"in ui" or "in frontend"

**Review by pallost — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 15 files reviewed, 8 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/components/routes-and-lines/line-details/ActionsRow.tsx` line 40 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8StRsA5DqP0s38p7VV:-N8StRsA5DqP0s38p7VW:b-s0bulc) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-details/ActionsRow.tsx#L40)):*
> ```XML
>             value={observationDate?.toISODate() || ''}
>             onChange={(e) =>
>               setObservationDate(DateTime.fromISO(e.target.value))
> ```

let's not directly write to the url query from an inline function, it will be hard to follow... Let's rather wrap this into an `onDateChange` function as it was before
___
*[`ui/src/components/routes-and-lines/line-details/RouteStopsSection.tsx` line 21 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SlMpN0ZmproUJr-lk:-N8SlMpN0ZmproUJr-ll:bg7op39) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-details/RouteStopsSection.tsx#L21)):*
> ```XML
>   className?: string;
>   route: RouteRoute;
>   observationDate?: DateTime;
> ```

let's not allow undefined here. The useGetLineDetails should anyway always default to the current date for observationDate, so this is always set
___
*[`ui/src/components/routes-and-lines/line-details/RouteStopsTable.tsx` line 13 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SlbonEZG4mTCbwPYJ:-N8SlbonEZG4mTCbwPYK:bg7op39) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-details/RouteStopsTable.tsx#L13)):*
> ```XML
>   className?: string;
>   routes: RouteRoute[];
>   observationDate?: DateTime;
> ```

let's not allow undefined here. The useGetLineDetails should anyway always default to the current date for observationDate, so this is always set
___
*[`ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx` line 30 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SpzsYEQHgXxwW9qJ6:-N8Su-4f2wPYIwwYCO-D:b-h9vjrs) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx#L30)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

Why do we have to set "initial" observation date to query? IMO it is valid case that it is not there (and then we would fallback to e.g. today in the UI)
</blockquote></details>

multiple hooks and components are communicating the observationDate state through the query. In case of e.g. useGetLineDetails hook, this is a must as we cannot necessarily always fall back to "today" as default, as it might not be within the line's validity period. IMO, this is perfectly fine solution. (I suggested it ;) ) Though maybe I'd rather move this initialization logic directly to useGetLineDraftDetails hook, instead of having it here in the component
___
*[`ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx` line 63 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Sv9ea5T_Q3LZ6-5k0:-N8Sv9ea5T_Q3LZ6-5k1:b6xpsow) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx#L63)):*
> ```XML
>             value={observationDate?.toISODate() || ''}
>             onChange={(e) =>
>               setObservationDate(DateTime.fromISO(e.target.value))
> ```

instead of modifying the query directly from an inline function, let's wrap it to an event handler function for clarity and trackability
___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 42 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8St-CK92sBPbY033t9:-N8St-CK92sBPbY033tA:bkfj344) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L42)):*
> ```XML
>     () =>
>       parseDate(queryParams.observationDate as string | undefined) ??
>       DateTime.now(),
> ```

could you put a comment here that useMemo was used because DateTime.now is constantly changing?

**Review by pallost — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 15 files reviewed, 8 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/hooks/useUrlQuery.ts` line 6 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SqxSZ8JPV6Thv_GSy:-N8SwYb04L2S_QVTbiDW:b-8jwylb) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/useUrlQuery.ts#L6)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

Do we want to expand responsibilites of this hook or rather create new hook to do new things?

Seems like we could split this easily to 2 different hooks that would do just one thing each:
- `useUrlQuery()` (or readUrlQuery or parseUrlQuery or something) to do what this originally did
- `setToUrlQuery(newParam: Object)`: calls useUrlQuery and sets new query param based to it

Then we could have convenience hooks for setting different parts of url query (e.g. observationDate) to avoid having define same (string) names for parts of url query in different places. And then you could expose all the previous hooks from one general hook, so that UI components could access all of those from one import and from one useSomeHok call.
</blockquote></details>

Similar thoughts... I could imagine extending this hook with some generic functions to serialize and deserialize complex objects, like DateTimes, but the actual observationDate handling should be in its own hook. Maybe just baked into the useGetLineDetails and useGetLineDraftDetails hooks...?

**Review by pallost — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 15 files reviewed, 10 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 58 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SzLHaDrsq4csGIUHQ:-N8SzLHaDrsq4csGIUHR:bwknbd7) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L58)):*
> ```XML
>   // Filter routes by observationDate in front end (and not in gql query) to avoid
>   // unnecessary graphql queries which would cause the list to reload on every date change
>   const filteredRoutes = filterRoutesByObservationDate(routes, observationDate);
> ```

a bit of functional programming thinking: instead of writing a wrapper function like `filterRoutesByObservationDate`, it would be more reusable if you wrote the checker function separately, like this:

```
const isRouteActiveOnObservationDate = (route: RouteRoute, observationDate: DateTime) =>
  isDateInRange(observationDate, route.validity_start, route.validity_end);
```

and then write the filter like this:

```
const filteredRoutes = routes.filter((route) => isRouteActiveOnObservationDate(route, observationDate));
```

or even better (if you learn to read these):
```
const filteredRoutes = routes.filter(partialRight(isRouteActiveOnObservationDate, observationDate));
```
___
*[`ui/src/utils/gql.ts` line 52 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8T-Wmx1o_vAwh18W-e:-N8T-Wmy1myrqPngrfhO:b8i3jw1) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/utils/gql.ts#L52)):*
> ```XML
>     label: { _eq: label },
>   },
> });
> ```

why not reuse it? ;)

```
export const constructRouteLineLabelGqlFilter = (label: string) => ({
  route_line: constructLabelGqlFilter(label)
});

```

**Review by HenrikHartiala — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 15 files reviewed, 10 unresolved discussions (waiting on @pallost and @tk-vincit)

___
*[`ui/src/components/routes-and-lines/line-details/ActionsRow.tsx` line 40 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8StRsA5DqP0s38p7VV:-N8TACH28OwyQRgLp4fe:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-details/ActionsRow.tsx#L40)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

let's not directly write to the url query from an inline function, it will be hard to follow... Let's rather wrap this into an `onDateChange` function as it was before
</blockquote></details>

Done.
___
*[`ui/src/components/routes-and-lines/line-details/RouteStopsSection.tsx` line 21 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SlMpN0ZmproUJr-lk:-N8TAedl57Lk2ZbXFWMx:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-details/RouteStopsSection.tsx#L21)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

let's not allow undefined here. The useGetLineDetails should anyway always default to the current date for observationDate, so this is always set
</blockquote></details>

Done.
___
*[`ui/src/components/routes-and-lines/line-details/RouteStopsTable.tsx` line 13 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SlbonEZG4mTCbwPYJ:-N8TAfdC7TQVaDGILUNB:bs7wumx) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-details/RouteStopsTable.tsx#L13)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

let's not allow undefined here. The useGetLineDetails should anyway always default to the current date for observationDate, so this is always set
</blockquote></details>

Done. 
___
*[`ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx` line 30 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SpzsYEQHgXxwW9qJ6:-N8TamHf8WD24ZziALKj:b-spwz4d) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx#L30)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

multiple hooks and components are communicating the observationDate state through the query. In case of e.g. useGetLineDetails hook, this is a must as we cannot necessarily always fall back to "today" as default, as it might not be within the line's validity period. IMO, this is perfectly fine solution. (I suggested it ;) ) Though maybe I'd rather move this initialization logic directly to useGetLineDraftDetails hook, instead of having it here in the component
</blockquote></details>

Good point, no business being in the component.
___
*[`ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx` line 63 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Sv9ea5T_Q3LZ6-5k0:-N8TIJC88a0hTFiyjX6g:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/components/routes-and-lines/line-drafts/LineDraftsPage.tsx#L63)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

instead of modifying the query directly from an inline function, let's wrap it to an event handler function for clarity and trackability
</blockquote></details>

Done.
___
*[`ui/src/hooks/useUrlQuery.ts` line 6 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SqxSZ8JPV6Thv_GSy:-N8U2SvB5yTkB95uleRt:b-1q51de) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/useUrlQuery.ts#L6)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

Similar thoughts... I could imagine extending this hook with some generic functions to serialize and deserialize complex objects, like DateTimes, but the actual observationDate handling should be in its own hook. Maybe just baked into the useGetLineDetails and useGetLineDraftDetails hooks...?
</blockquote></details>

This was a great idea. I did retrace my steps and created this useSetToUrlQuery -hook and useSetObservationToUrl -hook. This took a while, but at least I now learned why I should've not extend the useUrlQuery and rather make a new one. We'll see if my solution is fine now.
___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 42 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8St-CK92sBPbY033t9:-N8TMOat5_W4clv7uKUG:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L42)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

could you put a comment here that useMemo was used because DateTime.now is constantly changing?
</blockquote></details>

Done.
___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 56 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SsjKbBMDChXhi0Vha:-N8TMSH63sDtq_TOwR3f:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L56)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

"in ui" or "in frontend"
</blockquote></details>

Done.
___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 58 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SzLHaDrsq4csGIUHQ:-N8TRgFu8aHR9hnhBscU:b-vides7) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L58)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

a bit of functional programming thinking: instead of writing a wrapper function like `filterRoutesByObservationDate`, it would be more reusable if you wrote the checker function separately, like this:

```
const isRouteActiveOnObservationDate = (route: RouteRoute, observationDate: DateTime) =>
  isDateInRange(observationDate, route.validity_start, route.validity_end);
```

and then write the filter like this:

```
const filteredRoutes = routes.filter((route) => isRouteActiveOnObservationDate(route, observationDate));
```

or even better (if you learn to read these):
```
const filteredRoutes = routes.filter(partialRight(isRouteActiveOnObservationDate, observationDate));
```
</blockquote></details>

Ah good one! I'll go with the first solution, since I think its more readable.
Also checked this `isDateInRange` function in `time.ts` and modified it to handle null startDate's as well. Will put this in a separate commit.
___
*[`ui/src/utils/gql.ts` line 52 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8T-Wmx1o_vAwh18W-e:-N8TSdCd70B5Fmqdj3eq:b3ax0it) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/utils/gql.ts#L52)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

why not reuse it? ;)

```
export const constructRouteLineLabelGqlFilter = (label: string) => ({
  route_line: constructLabelGqlFilter(label)
});

```
</blockquote></details>

Well, I think that it is a bit more readable in this simple case to write the whole structure on this function without using the function above. If it's 
```
export const constructRouteLineLabelGqlFilter = (label: string) => ({
  route_line: constructLabelGqlFilter(label),
});
```
Then you would need to check what does this `constructLabelGqlFilter(label)` do, and so on. On the other hand, it would be more clear what is the difference between these two functions, so actually I'll do the change after all.

**Review by tk-vincit — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 20 files reviewed, 9 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/hooks/useUrlQuery.ts` line 6 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8SqxSZ8JPV6Thv_GSy:-N8XJoNSDy_GSwsbrpun:b-7smowg) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/hooks/useUrlQuery.ts#L6)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

This was a great idea. I did retrace my steps and created this useSetToUrlQuery -hook and useSetObservationToUrl -hook. This took a while, but at least I now learned why I should've not extend the useUrlQuery and rather make a new one. We'll see if my solution is fine now.
</blockquote></details>

IMO this is fine now.

**Review by pallost — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 20 files reviewed, 4 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/components/routes-and-lines/line-details/LineDetailsPage.tsx` line 79 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Y8tKdCKVOtGnSoZeC:-N8Y8tKdCKVOtGnSoZeD:b-31h3fb) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/components/routes-and-lines/line-details/LineDetailsPage.tsx#L79)):*
> ```XML
>                   <RouteStopsTable
>                     routes={line.line_routes}
>                     observationDate={observationDate ?? DateTime.now()}
> ```

the useGetLineDetails should always return a valid value for observationDate. If it's undefined in the beginning, you could extend the render condition to:

```
                {line.line_routes?.length > 0 && observationDate ? (
                  <RouteStopsTable
                    routes={line.line_routes}
                    observationDate={observationDate}
                ...
```
___
*[`ui/src/hooks/useSetObservationDateToUrl.ts` line 3 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8YA_e3ExDPro_K5Kq6:-N8YA_e3ExDPro_K5Kq7:bhmykcr) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/hooks/useSetObservationDateToUrl.ts#L3)):*
> ```XML
> import { useSetToUrlQuery } from './useSetToUrlQuery';
> 
> export const useSetObservationDateToUrl = () => {
> ```

umm, I was probably unclear about the responsibilities of the hooks.
- `useUrlQuery` should be responsible for generally reading and writing to the query, without actually knowing what parameters there are and what they mean
- `useObservationDateQueryParam` (so this file) should be responsible for reading and writing the observationDate parameter to the url. To give a meaning to the parameter and additionally do some type conversion. It should look roughly like this:

```
export const useObservationDateQueryParam = () => {
  const queryParams = useUrlQuery();
  const observationDate = parseDate(queryParams['observationDate']);

  const setObservationDate = (date: string, replace = false) => {
    setToUrlQuery({ paramName: 'observationDate', value: date, replace });
  };

  return {observationDate, setObservationDate}
}
```
___
*[`ui/src/hooks/useSetToUrlQuery.ts` line 10 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Y9sqfC3BULx33OV2S:-N8Y9sqfC3BULx33OV2T:bbdittl) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/hooks/useSetToUrlQuery.ts#L10)):*
> ```XML
>   const history = useHistory();
> 
>   /** Sets parameter to URL query
> ```

thanks for the comment!

I think function should actually go rather to useUrlQuery, no need to create a separate hooks for reading from and writing to the query, they belong to the same utility
___
*[`ui/src/utils/gql.ts` line 52 at r3](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8T-Wmx1o_vAwh18W-e:-N8YAU8h2g7K9R7TaBf3:blsfwj9) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8696c700246efda7ff3ac50545a66a82830d62df/ui/src/utils/gql.ts#L52)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

Well, I think that it is a bit more readable in this simple case to write the whole structure on this function without using the function above. If it's 
```
export const constructRouteLineLabelGqlFilter = (label: string) => ({
  route_line: constructLabelGqlFilter(label),
});
```
Then you would need to check what does this `constructLabelGqlFilter(label)` do, and so on. On the other hand, it would be more clear what is the difference between these two functions, so actually I'll do the change after all.
</blockquote></details>

Up to you :)

**Review by HenrikHartiala — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 21 files reviewed, 4 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/components/routes-and-lines/line-details/LineDetailsPage.tsx` line 79 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Y8tKdCKVOtGnSoZeC:-N8YDaqJ8_AyxfUdxQ0X:b-80bnvp) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/components/routes-and-lines/line-details/LineDetailsPage.tsx#L79)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

the useGetLineDetails should always return a valid value for observationDate. If it's undefined in the beginning, you could extend the render condition to:

```
                {line.line_routes?.length > 0 && observationDate ? (
                  <RouteStopsTable
                    routes={line.line_routes}
                    observationDate={observationDate}
                ...
```

</blockquote></details>

True.
___
*[`ui/src/hooks/useSetToUrlQuery.ts` line 10 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Y9sqfC3BULx33OV2S:-N8YFjDU3pWZEDrIBAir:bf5og1o) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/hooks/useSetToUrlQuery.ts#L10)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

thanks for the comment!

I think function should actually go rather to useUrlQuery, no need to create a separate hooks for reading from and writing to the query, they belong to the same utility
</blockquote></details>

It was like that before, but as suggested by @tk-vincit I split them up in to two, which both now have a single responsibility. If we still decide that we want to combine these yet again, we do it in some other separate PR. 
___
*[`ui/src/hooks/useSetObservationDateToUrl.ts` line 3 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8YA_e3ExDPro_K5Kq6:-N8YFbusEcovQBspbQcx:bxd8bsx) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/hooks/useSetObservationDateToUrl.ts#L3)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

umm, I was probably unclear about the responsibilities of the hooks.
- `useUrlQuery` should be responsible for generally reading and writing to the query, without actually knowing what parameters there are and what they mean
- `useObservationDateQueryParam` (so this file) should be responsible for reading and writing the observationDate parameter to the url. To give a meaning to the parameter and additionally do some type conversion. It should look roughly like this:

```
export const useObservationDateQueryParam = () => {
  const queryParams = useUrlQuery();
  const observationDate = parseDate(queryParams['observationDate']);

  const setObservationDate = (date: string, replace = false) => {
    setToUrlQuery({ paramName: 'observationDate', value: date, replace });
  };

  return {observationDate, setObservationDate}
}
```
</blockquote></details>

Ah yeah, `useObservationDateQueryParm` with the parsing makes sense.  Thanks

**Review by pallost — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/hooks/useSetToUrlQuery.ts` line 10 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Y9sqfC3BULx33OV2S:-N8YWfwu-Sps4I032N7O:b-eblnqi) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/hooks/useSetToUrlQuery.ts#L10)):*
<details><summary><i>Previously, HenrikHartiala wrote…</i></summary><blockquote>

It was like that before, but as suggested by @tk-vincit I split them up in to two, which both now have a single responsibility. If we still decide that we want to combine these yet again, we do it in some other separate PR. 
</blockquote></details>

we just discussed it with @tk-vincit in the office and probably still having issues communicating the idea:
useUrlQuery should be responsible for generally reading and writing query parameters
useObservationDateQueryParam should give a meaning to a set of query parameters (in this case, 'observationDate') and do some additional conversion/parsing to it.

there's no need to split useUrlQuery and useSetToUrlQuery, they are both responsible for providing a generic interface to getting and setting parameters to the query

**Review by HenrikHartiala — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/hooks/useSetToUrlQuery.ts` line 10 at r7](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8Y9sqfC3BULx33OV2S:-N8Ym3bR9qUsW5ntkL3s:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/8dceb6d9ba1c97ba7075e01665c8f969f8f6596a/ui/src/hooks/useSetToUrlQuery.ts#L10)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

we just discussed it with @tk-vincit in the office and probably still having issues communicating the idea:
useUrlQuery should be responsible for generally reading and writing query parameters
useObservationDateQueryParam should give a meaning to a set of query parameters (in this case, 'observationDate') and do some additional conversion/parsing to it.

there's no need to split useUrlQuery and useSetToUrlQuery, they are both responsible for providing a generic interface to getting and setting parameters to the query
</blockquote></details>

Done.

**Review by pallost — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @HenrikHartiala, @pallost, and @tk-vincit)

___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 30 at r13](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8bLAcVAOvqCwucjOkC:-N8bLAcVAOvqCwucjOkD:b-h4999x) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/15883997dde84fde8482e807208edd1544e38c50/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L30)):*
> ```XML
>     useObservationDateQueryParam();
> 
>   // useMemo hook is used because DateTime.now() is constantly changing
> ```

this comment was forgotten here

**Review by HenrikHartiala — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/267)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @pallost and @tk-vincit)

___
*[`ui/src/hooks/line-drafts/useGetLineDraftDetails.ts` line 30 at r13](https://reviewable.io/reviews/hsldevcom/jore4-ui/267#-N8bLAcVAOvqCwucjOkC:-N8bdggj-9fr5id6zvlR:b-ucfknh) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/15883997dde84fde8482e807208edd1544e38c50/ui/src/hooks/line-drafts/useGetLineDraftDetails.ts#L30)):*
<details><summary><i>Previously, pallost (Tamas Pallos) wrote…</i></summary><blockquote>

this comment was forgotten here
</blockquote></details>

Thanks!
